src/GovUk.Education.ExploreEducationStatistics.Admin/Services/UpdateChartFilesService.cs

```diff
@@ -0,0 +1,168 @@
+{
+    public class UpdateChartFilesService : IUpdateChartFilesService
+    {
+        private readonly ContentDbContext _contentDbContext;
+        private readonly ILogger _logger;
+
+        private readonly string _storageConnectionString;
+        
+        public UpdateChartFilesService(
+            IConfiguration configuration,
+            ContentDbContext contentDbContext,
+            ILogger<UpdateChartFilesService> logger)
+        {
+            _storageConnectionString = configuration.GetValue<string>("CoreStorage");
+            _contentDbContext = contentDbContext;
+            _logger = logger;
+        }
+
+        public async Task<Either<ActionResult, bool>> UpdateChartFiles()
+        {
+            var refs = _contentDbContext
+                .ReleaseFileReferences
+                .Where(rfr => rfr.ReleaseFileType == ReleaseFileTypes.Chart)
+                .ToList();
+
+            var result = await UpdateChartFilesAsync(refs);
+
+            if (result.IsLeft)
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        private async Task<Either<ActionResult, bool>> UpdateChartFilesAsync(List<ReleaseFileReference> chartReferences)
+        {
+            var blobContainer = await GetCloudBlobContainer();
+
+            foreach (var fileReference in chartReferences)
+            {
+                var oldName = fileReference.Filename;
+                var oldBlobPath = AdminReleasePathWithFilename(fileReference.ReleaseId, oldName);
+
+                // Files should exists first time this is run in storage
+                if (blobContainer.GetBlockBlobReference(oldBlobPath).Exists())
+                {
+                    if (await CopyBlob(fileReference, blobContainer, oldBlobPath, fileReference.Id))
+                    {
+                        // Change ref to any datablocks with charts where used
+
+                        await UpdateDataBlock(fileReference, fileReference.Id);
+
+                        await _contentDbContext.SaveChangesAsync();
+
+                        _logger.LogInformation(
+                            $"Update blob: {fileReference.ReleaseId}:{fileReference.Filename} successfully");
+                    }
+                }
+            }
+
+            return true;
+        }
+
+        private async Task<bool> CopyBlob(ReleaseFileReference fileReference, CloudBlobContainer blobContainer, string oldBlobPath, Guid newId)
+        {
+            var newBlobPath = AdminReleasePathWithFilename(fileReference.ReleaseId, newId.ToString());
+
+            var sourceBlob = blobContainer.GetBlockBlobReference(oldBlobPath);
+            var targetBlob = blobContainer.GetBlockBlobReference(newBlobPath);
+                        
+            await targetBlob.StartCopyAsync(sourceBlob);
+            targetBlob.FetchAttributes();
+            while (targetBlob.CopyState.Status == CopyStatus.Pending) {
```

Comment:
Wow, we can't just `await` on a copy method? This is a pretty crappy API :vomiting_face:

Reply:
Yeah surprised me as well!